diff --git a/build_core/conf/openwrt/openwrt/SConscript b/build_core/conf/openwrt/openwrt/SConscript
index c01a422..a18a5d1 100644
--- a/build_core/conf/openwrt/openwrt/SConscript
+++ b/build_core/conf/openwrt/openwrt/SConscript
@@ -28,6 +28,9 @@ vars.Add('TARGET_CXX', 'OpenWRT C++ compiler', os.environ.get('TARGET_CXX'))
 vars.Add('TARGET_LINK', 'OpenWRT Linker', os.environ.get('TARGET_CC')) # Normally use the C compiler for linking.
 vars.Add('TARGET_LINKFLAGS', 'OpenWRT Linker flags', os.environ.get('TARGET_LDFLAGS'))
 
+vars.Add('TARGET_AR', 'OpenWRT Archiver', os.environ.get('TARGET_AR'))
+vars.Add('TARGET_RANLIB', 'OpenWRT Archive Indexer', os.environ.get('TARGET_RANLIB'))
+
 vars.Add('STAGING_DIR', 'OpenWRT staging dir', os.environ.get('STAGING_DIR'))
 
 vars.Update(env)
@@ -53,6 +56,8 @@ env.MergeFlags(flags)
 env.Replace(CC = env['TARGET_CC'])
 env.Replace(CXX = env['TARGET_CXX'])
 env.Replace(LINK = env['TARGET_LINK'])
+env.Replace(AR = env['TARGET_AR'])
+env.Replace(RANLIB = env['TARGET_RANLIB'])
 
 env['ENV']['PATH'] = env['TARGET_PATH']
 env['ENV']['STAGING_DIR'] = env['STAGING_DIR']
