diff --git a/common/os/posix/IfConfigLinux.cc b/common/os/posix/IfConfigLinux.cc
index c01a422..d4462e1 100644
--- a/common/os/posix/IfConfigLinux.cc
+++ b/common/os/posix/IfConfigLinux.cc
@@ -343,7 +343,7 @@ static std::list<IfEntry> NetlinkGetInte
     // everyone else does and just allocate a  "big enough' buffer.  We choose
     // 64K which should handle about 150 interfaces.
     //
-    const uint32_t BUFSIZE = 65536;
+    const uint32_t BUFSIZE = 262144; /* 256K */
     char* buffer = new char[BUFSIZE];
     uint32_t len;
 
