--- a/drivers/mtd/nand/pxa3xx_nand.c
+++ b/drivers/mtd/nand/pxa3xx_nand.c
@@ -1755,7 +1776,11 @@ static int pxa3xx_nand_probe(struct plat
 		 * parameter. This name cannot be changed or otherwise
 		 * user's mtd partitions configuration would get broken.
 		 */
+#ifdef CONFIG_MACH_ARMADA_MAMBA
+		mtd->name = "armada-nand";
+#else
 		mtd->name = "pxa3xx_nand-0";
+#endif
 		info->cs = cs;
 		ret = pxa3xx_nand_scan(mtd);
 		if (ret) {
