diff --git a/arch/arm/boot/dts/armada-xp-mamba.dts b/arch/arm/boot/dts/armada-xp-mamba.dts
index d9b32a2..ef8e7b0 100644
--- a/arch/arm/boot/dts/armada-xp-mamba.dts
+++ b/arch/arm/boot/dts/armada-xp-mamba.dts
@@ -59,8 +59,8 @@
 
 		internal-regs {
 			pinctrl {
-				pinctrl-0 = <&pmx_phy_int>;
-				pinctrl-names = "default";
+				/*pinctrl-0 = <&pmx_keys>;*/
+				/*pinctrl-names = "default";*/
 
 				pmx_ge0: pmx-ge0 {
 					marvell,pins = "mpp0", "mpp1", "mpp2", "mpp3",
@@ -77,7 +77,7 @@
 				};
 
 				pmx_keys: pmx-keys {
-					marvell,pins = "mpp33";
+					marvell,pins = "mpp32", "mpp33";
 					marvell,function = "gpio";
 				};
 
@@ -85,11 +85,6 @@
 					marvell,pins = "mpp36", "mpp37", "mpp38", "mpp39";
 					marvell,function = "spi";
 				};
-
-				pmx_phy_int: pmx-phy-int {
-					marvell,pins = "mpp32";
-					marvell,function = "gpio";
-				};
 			};
 
 			serial@12000 {
@@ -166,8 +161,14 @@
 		pinctrl-names = "default";
 
 		button@1 {
+                        label = "WPS Button";
+                        linux,code = <0x211>; /* WPS_SETUP */
+                        gpios = <&gpio1 0 1>;
+                };
+
+		button@2 {
 			label = "Factory Reset Button";
-			linux,code = <141>; /* KEY_SETUP */
+			linux,code = <0x198>; /* KEY_SETUP */
 			gpios = <&gpio1 1 1>;
 		};
 	};
